host_core/lib/host_core/vhost/virtual_host.ex

```diff
@@ -0,0 +1,549 @@
+
+  defmodule State do
+    @type t :: %State{
+            config: Configuration.t(),
+            friendly_name: String.t(),
+            labels: Map.t(),
+            start_time: integer(),
+            supplemental_config: Map.t()
+          }
+    defstruct [:config, :friendly_name, :start_time, :labels, :supplemental_config]
+  end
+
+  @doc """
+  Starts the virtual host
+  """
+  @spec start_link(config :: Configuration.t()) :: :ignore | {:error, any} | {:ok, pid}
+  def start_link(config) do
+    GenServer.start_link(__MODULE__, config,
+      name: via_tuple(config.host_key, config.lattice_prefix)
+    )
+  end
+
+  @impl true
+  @spec init(config :: Configuration.t()) ::
+          {:ok, State.t()} | {:ok, State.t(), {:continue, atom()}}
+  def init(config) do
+    Process.flag(:trap_exit, true)
+
+    case HostCore.Lattice.LatticeRoot.start_lattice(config) do
+      {:ok, _pid} ->
+        Logger.info("Lattice supervisor #{config.lattice_prefix} started.")
+
+      {:error, e} ->
+        Logger.error("Lattice supervisor #{config.lattice_prefix} failed to start: #{inspect(e)}")
+    end
+
+    friendly_name = HostCore.Namegen.generate()
+
+    Logger.info("Virtual Host #{config.host_key} (#{friendly_name}) started.")
+    Logger.info("Virtual Host Issuer Public key: #{config.cluster_key}")
+    Logger.info("Valid cluster signers for host: #{config.cluster_issuers}")
+
+    if config.cluster_adhoc do
+      warning = """
+      WARNING. You are using an ad hoc generated cluster seed.
+      For any other host or CLI tool to communicate with this host,
+      you MUST copy the following seed key and use it as the value
+      of the WASMCLOUD_CLUSTER_SEED environment variable:
+
+      #{config.cluster_seed}
+
+      You must also ensure the following cluster signer is in the list of valid
+      signers for any new host you start:
+
+      #{config.cluster_issuers |> Enum.at(0)}
+
+      """
+
+      Logger.warn(warning)
+    end
+
+    labels =
+      get_env_host_labels()
+      |> Map.merge(HostCore.WasmCloud.Native.detect_core_host_labels())
+
+    Gnat.ConsumerSupervisor.start_link(%{
+      connection_name: HostCore.Nats.control_connection(config.lattice_prefix),
+      module: HostCore.ControlInterface.HostServer,
+      subscription_topics: [
+        %{
+          topic: "#{config.ctl_topic_prefix}.#{config.lattice_prefix}.cmd.#{config.host_key}.*"
+        },
+        %{
+          topic: "#{config.ctl_topic_prefix}.#{config.lattice_prefix}.get.#{config.host_key}.inv"
+        }
+      ]
+    })
+
+    {wclock, _} = :erlang.statistics(:wall_clock)
+
+    Logger.metadata(host_id: config.host_key, lattice_prefix: config.lattice_prefix)
+
+    state = %State{
+      config: Map.put(config, :labels, labels),
+      friendly_name: friendly_name,
+      start_time: wclock,
+      labels: labels
+    }
+
+    :timer.send_interval(@thirty_seconds, self(), :publish_heartbeat)
+    Process.send_after(self(), :publish_started, 500)
```

Comment:
Is there a reason we do this with a send_after instead of just doing it in the last handle_continue? I can't remember if we access data or something that makes this necessary to happen _after_ the whole init chain finishes

Reply:
I think this is leftover from when I was knee deep in the middle of refactoring and I'd run into a race condition that I blamed on the startup sequence. Refactored to use handle continue now.